src/lib.rs

```diff
@@ -158,53 +331,148 @@ impl<T> Quadtree<T> {
     }
 
     // recursive function
-    fn search<'a: 'b, 'b>(&'a self, distances: &mut Vec<(&'b T, f64)>, query_point: &Coordinate) {
-        if !self.boundary.contains(query_point) {
-            return;
+    fn search<'a>(
+        &'a self,
+        nearest_neighbers: &mut BinaryHeap<(NonNanFloat, &'a T)>,
+        query_point: &Coordinate,
+        k: usize,
+    ) {
+        if nearest_neighbers.len() == k {
+            if let Some((max_distance, _)) = nearest_neighbers.peek() {
+                if self.boundary.distance(query_point) >= *max_distance {
+                    return;
+                }
+            }
         }
 
-        if self.children.is_empty() {
-            for (coordinate, interest) in self.coordinates.iter().zip(self.interests.iter()) {
-                let distance = coordinate.distance(query_point);
-                distances.push((interest, distance));
-            }
-        } else {
-            for child in self.children.iter() {
-                child.search(distances, query_point);
+        for (coord, v) in zip(&self.coordinates, &self.interests) {
+            let distance = coord.distance(query_point);
+            if nearest_neighbers.len() < k {
+                nearest_neighbers.push((NonNanFloat::new(distance), v));
+            } else if let Some((max_distance, _)) = nearest_neighbers.peek() {
+                if distance >= **max_distance {
+                    continue;
+                }
+                if nearest_neighbers.len() == k {
+                    nearest_neighbers.pop();
+                }
+                nearest_neighbers.push((NonNanFloat::new(distance), v));
             }
         }
+
+        for child in self.children.iter() {
+            child.search(nearest_neighbers, query_point, k);
+        }
     }
 
     pub fn find_nearest_neighbors(&self, query_point: &Coordinate, k: usize) -> Vec<&T> {
-        let mut distances = vec![];
-
-        self.search(&mut distances, query_point);
-        distances.sort_by(|(_, dis1), (_, dis2)| dis1.partial_cmp(dis2).unwrap());
-        distances
-            .into_iter()
-            .take(k)
-            .map(|(c, _)| c)
-            .collect::<Vec<_>>()
+        let mut nearest_neighbors = BinaryHeap::new();
+        self.search(&mut nearest_neighbors, query_point, k);
+
+        nearest_neighbors
+            .into_sorted_vec()
```

Comment:
what would the cost of `into_sorted_vec` like? 

https://github.com/rust-lang/rust/issues/115357

Reply:
Considering the use case, it seems unlikely that we will need to use a large value for 'k'.